src/nanoemoji/svg.py

```diff
@@ -0,0 +1,189 @@
+from picosvg.svg_transform import Affine2D
+import regex
+from typing import MutableMapping, Sequence, Tuple
+
+
+def _create_doclist(svg: SVG) -> str:
+    return (
+        svg
+        # dumb sizing isn't useful
+        .remove_attributes(("width", "height"))
+        # Firefox likes to render blank if present
+        .remove_attributes(("enable-background",))
+    )
+
+
+def _ensure_has_id(el: etree.Element):
+    if "id" in el.attrib:
+        return
+    nth_child = 0
+    prev = el.getprevious()
+    while prev is not None:
+        nth_child += 1
+        prev = prev.getprevious()
+    el.attrib["id"] = f'{el.getparent().attrib["id"]}::{nth_child}'
+
+
+def _glyph_groups(color_glyphs: Sequence[ColorGlyph]) -> Tuple[Tuple[str, ...]]:
+    """Find glyphs that need to be kept together by union find."""
+    # glyphs by reuse_key
+    glyphs = {}
+    reuse_groups = DisjointSet()
+    for color_glyph in color_glyphs:
+        reuse_groups.make_set(color_glyph.glyph_name)
+        for painted_layer in color_glyph.as_painted_layers():
+            # TODO what attributes should go into this key for SVG
+            reuse_key = _inter_glyph_reuse_key(painted_layer)
+            if reuse_key not in glyphs:
+                glyphs[reuse_key] = color_glyph.glyph_name
+            else:
+                reuse_groups.union(color_glyph.glyph_name, glyphs[reuse_key])
+
+    return reuse_groups.sorted()
+
+
+def _add_unique_gradients(
+    id_updates: MutableMapping[str, str],
+    svg_defs: etree.Element,
+    color_glyph: ColorGlyph,
+):
+    for gradient in color_glyph.picosvg.xpath("//svg:defs/*"):
+        gradient = copy.deepcopy(gradient)
+        curr_id: str = gradient.attrib["id"]
+        new_id = f"{color_glyph.glyph_name}::{curr_id}"
+        del gradient.attrib["id"]
+        gradient_xml: str = etree.tostring(gradient)
+        if gradient_xml in id_updates:
+            id_updates[curr_id] = id_updates[gradient_xml]
+        else:
+            gradient.attrib["id"] = new_id
+            id_updates[curr_id] = new_id
+            id_updates[gradient_xml] = new_id
+            svg_defs.append(gradient)
+
+
+def _inter_glyph_reuse_key(painted_layer: PaintedLayer):
+    """Individual glyf entries, including composites, can be reused.
+
+    SVG reuses w/paint so paint is part of key."""
+
+    # TODO we could recycle shapes that differ only in paint, would just need to
+    # transfer the paint attributes onto the use element if they differ
+    return (painted_layer.paint, painted_layer.path.d, painted_layer.reuses)
+
+
+def _add_glyph(svg, color_glyph, id_updates):
+    # each glyph gets a group of its very own
+    svg_g = svg.append_to("/svg:svg", etree.Element("g"))
+    svg_g.attrib["id"] = color_glyph.glyph_name
+
+    # copy the shapes into our svg
+    glyphs = {}
+    for painted_layer in color_glyph.as_painted_layers():
+        reuse_key = _inter_glyph_reuse_key(painted_layer)
+        if reuse_key not in glyphs:
```

Comment:
You are resetting the `glyphs` dict every time you call `_add_glyph`, instead of initializing it in the caller frame, and passing it down and accummulating in `_add_glyph`.
So this is not actually sharing/reusing layers with same path/paint _between_ color_glyphs, only _within_ each color_glyph -- am I correct?
If so, the use of a function named `_inter_glyph_reuse_key` (instead of _intra_) is kind of misleading..

But should we not also share layers between glyphs that belong to the same disjoint set? I thought that was the point of grouping them in `_glyph_groups`, but I might be missing something very basic.. 🤔

Reply:
Ugh no that's a bug I caused when teasing apart an overlong function. For bonus points I have a test that specifically looks for this and "corrected" it's contents w/o noticing this had broken.

Reply:
Fixed. It's unfortunately fairly easy to miss what's changing in the test ttx because the svg content is not pretty printed.